### ISSUE TITLE
Cannot read property 'on' of undefined

### ISSUE THREAD

**Issue by ksnyde** [simplabs/ember-simple-auth#635]
I have installed the **jjabrams** ESA and wanted to get that in place before actually wiring it up to my OAuth based backend. With the use of the demo-app it wasn't too hard but I am getting two extraneous errors in my unit tests that appear to map back to ESA, they're both similar code blocks:

> #application-route-mixin.js

    _mapSessionEventsToActions: Ember.on('init', function() {
        Ember.A([
          ['authenticationSucceeded', 'sessionAuthenticated'],
          ['invalidationSucceeded', 'sessionInvalidated']
        ]).forEach(([event, method]) => {
          this.get('session').on(event, Ember.run.bind(this, () => {
            this[method](...arguments);
          }));
        });
      }),

> #addon/services/session.js

    _forwardSessionEvents: on('init', function() {
      Ember.A([
        'authenticationSucceeded',
        'invalidationSucceeded'
      ]).forEach((event) => {
        this.get('session').on(event, () => {
          this.trigger(event, ...arguments);
        });
      });
    }),

In both cases, the `this.get('session')` seems to resolve to an undefined object. In a project with 200 or so unit tests I only get one failure for each. The first one comes from the `route/application.js` where I've mixed in the `ApplicationRouteMixin` just like your demo-app. The other one is a bit more bizarre considering the error comes from a "it renders" test of a Component which *does not* inject the **session** service.

**Comment by rlivsey** [comment]
Have you injected the service into the route?

```javascript
session: Ember.inject.service()
```

The session used to be injected into everything automatically, but now needs to be done manually (to be in line with other services)

**Comment by ksnyde** [comment]
I haven't injected anywhere yet. Does it need to be injected in certain places? 

My cunning strategy was to get this in place without deprecations and then come back and wire it up once I had all other deprecations removed.

**Comment by rlivsey** [comment]
Anywhere you're accessing `this.get("session")` you'll need to have injected it into that route/controller.

**Comment by rlivsey** [comment]
… although if you're using the mixins they [do that for you](https://github.com/simplabs/ember-simple-auth/blob/jj-abrams/addon/mixins/application-route-mixin.js#L43)

**Comment by ksnyde** [comment]
yes but I'm never accessing session ... it's just the ESA mixin code. :)

**Comment by rlivsey** [comment]
Yeah I got myself confused, I mis-read and thought that code you pasted in was from your app!

**Comment by ksnyde** [comment]
So for instance my `route/application.js` is just:

    import Ember from 'ember';
      import ApplicationRouteMixin from 'ember-simple-auth/mixins/application-route-mixin';

      export default Ember.Route.extend(ApplicationRouteMixin,{
        session: Ember.inject.service()
      });

where I didn't have the session injection initially -- because the demo app does not -- but I thought maybe it somehow needs it. In both cases though I get the same error.

**Comment by marcoow** [comment]
You don't have to mixin the session service manually as the mixins should do that for you. Did you `git clean -dfx`?

**Comment by ksnyde** [comment]
Yeah the service addition was desperation. :8ball: 

WRT to `git clean` ... I hadn't considered doing this although I did clear out `node_modules` and `bower_modules` recently as part of the CLI upgrade. 

Just tried it and still getting the same results.

**Comment by marcoow** [comment]
please provide some more info - `package.json`, `bower.json`, ESA revision, `config/environment.js`…

**Comment by ksnyde** [comment]
here's all three: [gist](https://gist.github.com/ksnyde/700fdcfb5c0442201e7e)

**Comment by marcoow** [comment]
Run `npm update ember-simple-auth` to make suer you have the latest code.

**Comment by ssendev** [comment]
```coffee
`import Session from 'ember-simple-auth/session'`
`import EphemeralStore from 'ember-simple-auth/stores/ephemeral'`

moduleFor 'route:application', 'ApplicationRoute',
   needs: ['service:session']
   setup: ->
     @container.register 'session:main', Session
     @container.injection 'service:session', 'session', 'session:main'
     @container.register 'session-store:ephemeral', EphemeralStore
     @registry.injection 'session:main', 'store', 'session-store:ephemeral'
```
worked for me. it is needed because the session service currently isn't fully migrated to a service and internally still depends on the old session.

**Comment by marcoow** [comment]
@ssendev: the session service is fully migrated to a service - the dependency on the old session isn't going away.

**Comment by marcoow** [comment]
This should be fixed in the latest version of the `jj-abrams` branch - please reopen if the problem persists.

**Comment by bobisjan** [comment]
Hi,

I've just encountered this issue, but I'm trying to setup integration test for component like this:

```javascript
import { moduleForComponent, test } from 'ember-qunit';
import hbs from 'htmlbars-inline-precompile';
import simpleAuthInitializer from '../../initializers/ember-simple-auth';

moduleForComponent('some-component', 'Integration | Component | some component', {
  integration: true,
  setup() {
    simpleAuthInitializer.initialize(this.container);
  }
});

test('it renders', function(assert) {
  assert.expect(1);
  this.render(hbs`{{some-component}}`);
  assert.ok(true);
});
```

**Comment by ibroadfo** [comment]
I'm hitting this when trying to integration test components: a minimal test case is `ember g component-test main-navigation` in the dummy app. :smile: So the component has its session injected correctly.

What setup is required for an integration test vs unit test?

![screen shot 2015-09-24 at 14 39 31](https://cloud.githubusercontent.com/assets/1550/10074841/2d652f7a-62ca-11e5-859c-0a997e0cf2de.png)

**Comment by marcoow** [comment]
@ibroadfo: check the acceptance tests in the dummy app

**Comment by ibroadfo** [comment]
the ESA helpers seem to require an actual application to have been started, but this doesn't occur in component integration tests.

it's possible that integration tests are the wrong level to be trying to test auth-related stuff on, i guess?

**Comment by marcoow** [comment]
The test helpers are for acceptance tests really. For other kinds of tests I'd probably stub the session.

**Comment by bttf** [comment]
This issue still occurs on a vanilla ember-cli app when adding the `ApplicationRouteMixin` to application route; on latest jj-abrams branch.

**Comment by marcoow** [comment]
@bttf: please provide more info on what's actually failing as well as your `package.json`, `bower.json?, `config/environment.js`.

**Comment by bttf** [comment]
```
not ok 7 PhantomJS 1.9 - Unit | Route | application: it exists
    ---
        actual: >
            null
        message: >
            Died on test #1     at http://localhost:7357/assets/test-support.js:2934
                at test (http://localhost:7357/assets/test-support.js:1826)
                at http://localhost:7357/assets/shiiite.js:379
                at http://localhost:7357/assets/vendor.js:150
                at tryFinally (http://localhost:7357/assets/vendor.js:30)
                at http://localhost:7357/assets/vendor.js:156
                at http://localhost:7357/assets/test-loader.js:29
                at http://localhost:7357/assets/test-loader.js:21
                at http://localhost:7357/assets/test-loader.js:40
                at http://localhost:7357/assets/test-support.js:6656: 'undefined' is not an object (evaluating '_this.get('session').on')
        Log: |
    ...
```

Gist here: https://gist.github.com/bttf/85e0fff9c06326ee5f07

Steps to reproduce:
1. `ember new test-app && cd test-app`
2. Install jj-abrams branch of ember-simple auth
3. Add `ApplicationRouteMixin` to application route
4. Run tests

Also tried with ember v2.1 with same results

**Comment by marcoow** [comment]
@bttf: add `needs ['service:session']` to the test. As stated above, the test helpers are really for integration tests not unit tests.

**Comment by bttf** [comment]
Oops, forgot to include the `needs` dependency. Did that though; still receiving the same error as above.

I understand about the test helpers but in this case I don't think they are being involved? Just trying to test a route that has ApplicationRouteMixin included.

**Comment by marcoow** [comment]
Ah sorry, right. This isn't actually about the test helpers. You probably needs `needs: ['service:session', 'session-store:local-storage']` or whatever store you have configured.

**Comment by marcoow** [comment]
That was actually a bug in the code - fixed in 9334d47ae431ebc0e2ae48fc8fe089558e795296

**Comment by bttf** [comment]
Fix confirmed
